### PR TITLE
feat(datepicker): datepickers in popups

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**
This is ready for the review now. There are still a number of corner cases that we should take care about but IMO we should get it to the user hands ASAP.

I will open small, focused issues for all the remaining items as soon as this one lands.

### CODE REVIEW

File: src/datepicker/datepicker-input.ts [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+  /**
+   * @internal
+   */
+  manualDateChange(value: string) {
+    this._model = this._parserFormatter.parse(value);
+    this._onChange({year: this._model.year, month: this._model.month, day: this._model.day});
+    this._writeModelValue(this._model);
+  }
+
+  isOpen() { return !!this._cRef; }
```
Comment (wesleycho) [inline src/datepicker/datepicker-input.ts]: This would probably be better as a getter

File: src/datepicker/datepicker-input.spec.ts [hunk 2]
```diff
@@ -0,0 +1,220 @@
+import {TestBed, ComponentFixture, fakeAsync, tick} from '@angular/core/testing';
+import {By} from '@angular/platform-browser';
+import {createGenericTestComponent} from '../test/common';
+
+import {Component} from '@angular/core';
+import {FormsModule} from '@angular/forms';
+
+import {NgbDatepickerModule} from './datepicker.module';
+import {NgbInputDatepicker} from './datepicker-input';
```
Comment (jnizet) [inline src/datepicker/datepicker-input.spec.ts]: Shouldn't the directive be named NgbInputDatepicker, or the file be named input-datepicker?

Comment (wesleycho) [inline src/datepicker/datepicker-input.spec.ts]: I'm fine with this - the regular datepicker is `NgbDatepicker`, we namespace all of these for clarity that these are third party components & it creates a nice user cognition that they should be prefixing the ngb directives/components.

File: src/datepicker/datepicker-input.ts [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+  /**
+   * First day of the week, 0=Sun, 1=Mon, etc.
+   */
+  @Input() firstDayOfWeek: number;
+
+  /**
+   * Callback to mark a given date as disabled
+   */
+  @Input() markDisabled: (date: {year: number, month: number, day: number}) => boolean;
```
Comment (jnizet) [inline src/datepicker/datepicker-input.ts]: Couldn't we create an interface for this structure. What would be the disadvantage of doing so? It would make our code easier to read (IMO), and the client code safer too, because it could use that interface too, and thus avoid typos like `seconds` instead of `second`?

Comment (wesleycho) [inline src/datepicker/datepicker-input.ts]: Isn't it basically a NgbDate?

File: src/datepicker/datepicker-input.ts [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+  /**
+   * Whether to display week numbers
+   */
+  @Input() showWeekNumbers: boolean;
+
+  /**
+   * Date to open calendar with. If nothing provided, calendar will open with current month.
+   * Use 'navigateTo(date)' as an alternative
+   */
+  @Input() startDate: {year: number, month: number};
```
Comment (jnizet) [inline src/datepicker/datepicker-input.ts]: I think this should be renamed to `startMonth` (and in datepicker, too).

File: src/datepicker/datepicker-input.ts (hunk 3 above)
Comment (pkozlowski-opensource) [inline src/datepicker/datepicker-input.ts]: Guys, this is not part of this PR, it is moved over from the NgbDatePicker. If you want to tackle this let's do it in a separate issue - this PR is a big one already. My quick take on it:
* yes, let's create an interface
* no, it is not the same as `NgbDate` - `NgbDate` is meant to be internal and shouldn't be exposed to the end users

File: src/datepicker/datepicker-input.ts [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+  }
+
+  close() {
+    if (this.isOpen()) {
+      this._vcRef.remove(this._vcRef.indexOf(this._cRef.hostView));
+      this._cRef = null;
+    }
+  }
+
+  toggle() {
```
Comment (jnizet) [inline src/datepicker/datepicker-input.ts]: These methods should be documented.

File: src/datepicker/datepicker-input.ts (hunk 4 above)
Comment (pkozlowski-opensource) [inline src/datepicker/datepicker-input.ts]: Once again, this is not really part of this PR. If this needs to be discussed please create a separate issue.

File: src/datepicker/datepicker-input.ts (hunk 5 above)
Comment (pkozlowski-opensource) [inline src/datepicker/datepicker-input.ts]: Oh yes! Good catch, thnx!

File: demo/src/app/components/datepicker/demos/popup/datepicker-popup.html [hunk 6]
```diff
@@ -0,0 +1,14 @@
+<form class="form-inline">
+  <div class="form-group">
+    <div class="input-group">
+      <input class="form-control" placeholder="yyyy-mm-dd"
+             name="dp" [(ngModel)]="model" ngbDatepicker #d="ngbDatepicker">
+      <div class="input-group-addon" (click)="d.toggle()" >
+        <img src="img/calendar-icon.svg" style="width: 1.2rem; height: 1.2rem; cursor: pointer;"/>
```
Comment (jnizet) [inline demo/src/app/components/datepicker/demos/popup/datepicker-popup.html]: Here, on Firefox and Chrome, the height of the addon is a bit too large. Setting the height to 1 rem makes it perfectly aligned with the input.

**Comment by jnizet** [comment]
This looks really great. Congrats. 👍 

I'm not sure if it's one of the known bugs that you plan to fix later, so I'll report it here. From the demo, open the popup, pick a date, erase the last two digits from the input field, click elsewhere in the page, then click the addon button again: an exception is thrown (`Cannot read property 'year' of null at NgbInputDatepicker.manualDateChange`), and it's not possible to open the popup anymore without refreshing the page.

**Comment by pkozlowski-opensource** [comment]
> This looks really great. Congrats.

thnx @jnizet 

> I'm not sure if it's one of the known bugs that you plan to fix later, so I'll report it here. From the demo, open the popup, pick a date, erase the last two digits from the input field, click elsewhere in the page, then click the addon button again: an exception is thrown (Cannot read property 'year' of null at NgbInputDatepicker.manualDateChange), and it's not possible to open the popup anymore without refreshing the page.

There are still many, many corner cases around invalid dates / inputs, validation and popup opening. I will fix the one you've found but there are many more tests we should add to make it really robust. In any case we are still in alpha so as long as the happy path works I want people to play with it.

But yeh, it definitively needs more polish.

File: demo/src/app/components/datepicker/demos/popup/datepicker-popup.html (hunk 6 above)
Comment (maxokorokov) [inline demo/src/app/components/datepicker/demos/popup/datepicker-popup.html]: I would say just remove height altogether and leave width 1.2rem

**Comment by maxokorokov** [comment]
Looks quite good for the initial version!
Sure the validation/conversion/focus/blur should be improved later.

One thing that needs to be fixed for sure for me is when you type something invalid like `aaa` in the input and try to open, it crashes the datepicker forever with `TypeError: Attempted to assign to readonly property.` To me the first version should eat all invalid inputs silently and probably avoid formatting `2016` to `2016--`

Another is that there is no documentation on the `NgbInputDatepicker` in the demo currently.

It's sad that we have all datepicker inputs in three places, but I can't see what to do with it right away...

File: src/datepicker/datepicker-input.ts (hunk 1 above)
Comment (pkozlowski-opensource) [inline src/datepicker/datepicker-input.ts]: `isOpen()` is consistent with what we've got in other directives (dropdown, typeahead, ...). Please open a separate issue if you think that we should change it across the project.

**Comment by pkozlowski-opensource** [comment]
@jnizet @maxokorokov @wesleycho once again thnx for all your reviews! I went through all the comments and fixed pbs or commented on your comments :-)

I've also fixed bugs you've seen - I'm sure that more remain but yeh, that's just a starting point.

**Comment by wesleycho** [comment]
I'm fine with this as it is for now.
